components/file_system/src/condvar.rs

```diff
@@ -0,0 +1,299 @@
+            CondvarNode::Async(_, ref node) => node.prev.set(prev),
+        }
+    }
+
+    pub fn set_next(&self, next: Option<NonNull<CondvarNode>>) {
+        match *self {
+            CondvarNode::Sync(_, ref node) => node.next.set(next),
+            CondvarNode::Async(_, ref node) => node.next.set(next),
+        }
+    }
+}
+
+/// Un-prioritized conditional variable. Supports both synchronously or
+/// asynchronously waiting on the same instance.
+/// TODO: Prioritized waiting
+/// Maintains multiple linked list for each priority. When notified, instead of
+/// waking up all nodes at once, sort them to a seperate waiting queue and only
+/// wake up the leader.
+#[derive(Debug)]
+pub struct Condvar {
+    head: Cell<Option<NonNull<CondvarNode>>>,
+    tail: Cell<Option<NonNull<CondvarNode>>>,
+}
+
+unsafe impl Send for Condvar {}
+unsafe impl Sync for Condvar {}
+
+impl Condvar {
+    pub fn new() -> Condvar {
+        Condvar {
+            head: Cell::new(None),
+            tail: Cell::new(None),
+        }
+    }
+
+    #[inline]
+    fn enqueue(&self, raw_node: &mut CondvarNode) {
+        let node = unsafe { Some(NonNull::new_unchecked(raw_node)) };
+        raw_node.set_prev(self.tail.get());
+        if let Some(tail) = self.tail.get() {
+            unsafe {
+                tail.as_ref().set_next(node);
+            }
+        } else {
+            self.head.set(node);
+        }
+        self.tail.set(node);
+    }
+
+    #[inline]
+    fn dequeue(&self, raw_node: &mut CondvarNode) {
+        let prev = raw_node.get_prev();
+        let next = raw_node.get_next();
+        if let Some(prev) = prev {
+            unsafe {
+                prev.as_ref().set_next(next);
+            }
+        } else {
+            assert!(self.head.get().unwrap().as_ptr() == raw_node);
+            self.head.set(next);
+        }
+        if let Some(next) = next {
+            unsafe {
+                next.as_ref().set_prev(prev);
+            }
+        } else {
+            assert!(self.tail.get().unwrap().as_ptr() == raw_node);
+            self.tail.set(prev);
+        }
+    }
+
+    /// Notifies all waiters in queue as till now.
+    pub fn notify_all(&self) {
+        let mut ptr = self.head.get();
+        while let Some(inner) = ptr {
+            unsafe {
+                let node = inner.as_ref();
+                node.notify();
+                ptr = node.get_next();
+            }
+        }
+    }
+
+    /// Synchronously waits on this condition variable for a notification,
+    /// timing out after a specified duration.
+    pub fn wait_timeout<'a, T>(
+        &self,
+        guard: MutexGuard<'a, T>,
+        timeout: Duration,
+    ) -> (MutexGuard<'a, T>, bool) {
+        // mutable just to indulge NonNull
+        let mut node = CondvarNode::new_sync();
+        self.enqueue(&mut node);
+        // alternative: std::thread::park_timeout suffers from spurious wake
```

Comment:
consider using https://docs.rs/parking_lot/0.11.1/parking_lot/struct.Condvar.html

Reply:
parking_lot::Condvar is used in combo with parking_lot::Mutex, I would rather our implementation of Condvar can integrate well with standard Mutex. Also, I think a fair portion of it's optimizations aim at multiple waiter on the same condvar, which isn't touched by our use case (linked wait).

Reply:
you can use parking_lot::Mutex instead of standard Mutex as parking_lot::Mutex is 1.5x faster than std::sync::Mutex when uncontended